sys-firmware/midisport-firmware/files/midisport-firmware-1.2-rules.patch

```diff
@@ -23,5 +23,4 @@
 -ACTION=="add", SUBSYSTEM=="usb", DEVPATH=="/*.0", ENV{PRODUCT}=="763/1031/121", RUN+="@fxload@ -s @firmwaredir@/MidiSportLoader.ihx -I @firmwaredir@/MidiSport8x8-2.21.ihx"
 -
 -# vim: ft=conf
-+#ACTION=="add", SUBSYSTEM=="usb", DEVPATH=="/*.0", ENV{PRODUCT}=="763/1031/110", RUN+="@fxload@ -s @firmwaredir@/MidiSportLoader.ihx -I @firmwaredir@/MidiSport8x8-2.10.ihx"
-+#ACTION=="add", SUBSYSTEM=="usb", DEVPATH=="/*.0", ENV{PRODUCT}=="763/1031/121", RUN+="@fxload@ -s @firmwaredir@/MidiSportLoader.ihx -I @firmwaredir@/MidiSport8x8-2.21.ihx"
++ACTION=="add", SUBSYSTEM=="usb*", ATTRS{idVendor}=="0763", ATTRS{idProduct}=="1031", RUN+="@fxload@ -s @firmwaredir@/MidiSportLoader.ihx -I @firmwaredir@/MidiSport8x8-2.21.ihx -D %N"
```

Comment:
Well, then:
```
>>> Emerging (1 of 1) sys-firmware/midisport-firmware-1.2-r1::testworld
>>> Downloading 'http://downloads.sourceforge.net/usb-midi-fw/midisport-firmware-1.2.tar.gz'
--2020-07-20 12:21:06--  http://downloads.sourceforge.net/usb-midi-fw/midisport-firmware-1.2.tar.gz
Resolving downloads.sourceforge.net... 216.105.38.13
Connecting to downloads.sourceforge.net|216.105.38.13|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: http://downloads.sourceforge.net/project/usb-midi-fw/midisport-firmware/1.2/midisport-firmware-1.2.tar.gz [following]
--2020-07-20 12:21:08--  http://downloads.sourceforge.net/project/usb-midi-fw/midisport-firmware/1.2/midisport-firmware-1.2.tar.gz
Reusing existing connection to downloads.sourceforge.net:80.
HTTP request sent, awaiting response... 302 Found
Location: https://deac-riga.dl.sourceforge.net/project/usb-midi-fw/midisport-firmware/1.2/midisport-firmware-1.2.tar.gz [following]
--2020-07-20 12:21:08--  https://deac-riga.dl.sourceforge.net/project/usb-midi-fw/midisport-firmware/1.2/midisport-firmware-1.2.tar.gz
Resolving deac-riga.dl.sourceforge.net... 89.111.52.100
Connecting to deac-riga.dl.sourceforge.net|89.111.52.100|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 62948 (61K) [application/x-gzip]
Saving to: '/var/tmp/distfiles/midisport-firmware-1.2.tar.gz.__download__'

/var/tmp/distfiles/ 100%[===================>]  61.47K  --.-KB/s    in 0.05s   

2020-07-20 12:21:09 (1.10 MB/s) - '/var/tmp/distfiles/midisport-firmware-1.2.tar.gz.__download__' saved [62948/62948]

 * midisport-firmware-1.2.tar.gz BLAKE2B SHA512 size ;-) ...                                 [ ok ]
 * Determining the location of the kernel source code
 * Found kernel source directory:
 *     /usr/src/linux
 * Found sources for kernel version:
 *     5.7.0-pf4
 * Checking for suitable kernel configuration options...
 *   CONFIG_SND_USB_AUDIO:	 is not set when it should be.
 * Please check to make sure these options are set correctly.
 * Failure to do so may cause unexpected problems.
>>> Unpacking source...
>>> Unpacking midisport-firmware-1.2.tar.gz to /var/tmp/portage/sys-firmware/midisport-firmware-1.2-r1/work
>>> Source unpacked in /var/tmp/portage/sys-firmware/midisport-firmware-1.2-r1/work
>>> Preparing source in /var/tmp/portage/sys-firmware/midisport-firmware-1.2-r1/work/midisport-firmware-1.2 ...
 * Applying midisport-firmware-1.2-configure.patch ...                                       [ ok ]
 * Applying midisport-firmware-1.2-rules.patch ...
patching file 42-midisport-firmware.rules.in
patch: **** malformed patch at line 26:  
                                                                                             [ !! ]
 * ERROR: sys-firmware/midisport-firmware-1.2-r1::testworld failed (prepare phase):
 *   patch -p1  failed with /var/tmp/portage/sys-firmware/midisport-firmware-1.2-r1/files/midisport-firmware-1.2-rules.patch
 * 
 * Call stack:
 *               ebuild.sh, line  125:  Called src_prepare
 *             environment, line 2055:  Called default
 *      phase-functions.sh, line  855:  Called default_src_prepare
 *      phase-functions.sh, line  920:  Called __eapi6_src_prepare
 *             environment, line  207:  Called eapply '/var/tmp/portage/sys-firmware/midisport-firmware-1.2-r1/files/midisport-firmware-1.2-configure.patch' '/var/tmp/portage/sys-firmware/midisport-firmware-1.2-r1/files/midisport-firmware-1.2-rules.patch'
 *             environment, line  918:  Called _eapply_patch '/var/tmp/portage/sys-firmware/midisport-firmware-1.2-r1/files/midisport-firmware-1.2-rules.patch'
 *             environment, line  856:  Called __helpers_die 'patch -p1  failed with /var/tmp/portage/sys-firmware/midisport-firmware-1.2-r1/files/midisport-firmware-1.2-rules.patch'
 *   isolated-functions.sh, line  112:  Called die
 * The specific snippet of code:
 *   		die "$@"
 * 
 * If you need support, post the output of `emerge --info '=sys-firmware/midisport-firmware-1.2-r1::testworld'`,
 * the complete build log and the output of `emerge -pqv '=sys-firmware/midisport-firmware-1.2-r1::testworld'`.
 * The complete build log is located at '/var/tmp/portage/sys-firmware/midisport-firmware-1.2-r1/temp/build.log'.
 * The ebuild environment file is located at '/var/tmp/portage/sys-firmware/midisport-firmware-1.2-r1/temp/environment'.
 * Working directory: '/var/tmp/portage/sys-firmware/midisport-firmware-1.2-r1/work/midisport-firmware-1.2'
 * S: '/var/tmp/portage/sys-firmware/midisport-firmware-1.2-r1/work/midisport-firmware-1.2'

>>> Failed to emerge sys-firmware/midisport-firmware-1.2-r1
```

Did it work for you?

Reply:
! That doesn't make sense. I'll figure it out. Teething issues! Sorry!

Reply:
@juippis - what was the exact command you ran ? perhaps my testing was incorrect and I ran the wrong thing.

Reply:
It was simply trying to `emerge` it after applying this PR to my local copy of ::gentoo tree. The error is because you've hand-edited the patch file, not created it properly. `@@ -1,18 +1,13 @@` should be different.

Reply:
That'd do it. I thought I had generated it properly, but you're right, I did edit it by hand, so I made a mistake.
When you say "not created it properly" - is there a canonical way of doing it properly  ? Gentoo's way of doing things is very solid and I'd like to adhere to those guidelines if possible.

Reply:
Just in general how you create patch files. We use `p1` level as probably everything else uses it nowadays, to be consistent among other distributions and OSS projects.

You can google any guide how to create patches, it's not that complicated. 
```
1: get source code, 
2: unpack it, 
3: create a copy, cp -r dir/ dir-patched/, 
4: browse to dir-patched/where/is/the/broken/source/file, 
5: edit it, 
6: cd back to where dir/ and dir-patched/ exists
7: diff -Naur dir/ dir-patched/ > mypatch.patch
```
But ideally we'd like the headers and author of a patch file to be present in the patch, you can get it by doing a git diff style patch. Remember to have package name and version in the patch name, `midisport-firmware-1.2-fix-udev-rules.patch`.

Remember to always test your contribution with `emerge` or at least `ebuild` prior to pushing, so you can still fix things if it's not working, before you push..

Reply:
Excellent - thank you - I'll make sure I follow that procedure. I'll put in a patch tonight and hopefully this review will look good tomorrow morning.

Reply:
@juippis - should be updated now

Reply:
@juippis - wanted to check to see if there was anything else that needed to be done - I think this should be good at this point. (sorry for the poke, if it's bad etiquette, I'll leave it alone.)